### PR TITLE
Make deprecation warnings compile errors

### PR DESCRIPTION
**Pull request by wilzbach**
This is already the way we build Phobos + Tools.
See also: https://github.com/dlang/phobos/pull/5546

CC @ZombineDev @CyberShadow

### CODE REVIEW

**Comment by CyberShadow** [comment]
Sounds good, currently fails CI though.

**Comment by wilzbach** [comment]
> Sounds good, currently fails CI though.

Yeah:

```
src/core/time.d(4581): Deprecation: struct core.time.FracSec is deprecated - FracSec has been deprecated in favor of just using Duration for the sake of simplicity
src/core/time.d(3594): Error: template instance core.time._assertThrown!(TimeException, FracSec) error instantiating
```

Easy to fix.

```
src/core/internal/arrayop.d(147): Error: static assert  "Unary `-` not supported for type `ubyte`."
```

That one turns out to be rather difficult as it appears deep in the `arrayop` mixin :/

**Comment by wilzbach** [comment]
Updated to resolve the merge conflicts and incorporate newer deprecations.
For `delete` https://github.com/dlang/dmd/pull/8016 is required, apart from that I still haven't figured out how to fix the deprecation in `arrayop`.

File: src/rt/tracegc.d [hunk 1]
```diff
@@ -40,14 +40,15 @@ version (none)
...
-        S* ps; delete ps;
+        // delete is deprecated
+        //int[] a; delete a;
+        //S[] as; delete as;
+        //C c; delete c;
+        //I i; delete i;
+        //C* pc = &c; delete *pc;
+        //I* pi = &i; delete *pi;
+        //int* pint; delete pint;
+        //S* ps; delete ps;
```
Comment (wilzbach) [inline src/rt/tracegc.d]: Note that this is in a `version(none)` block.

Comment (Geod24) [reply src/rt/tracegc.d]: Deprecation message issued at parse time ? :/

Comment (JinShil) [reply src/rt/tracegc.d]: `delete` has been deprecated for a year now.  It's time to turn it into an error.

Comment (jacob-carlborg) [reply src/rt/tracegc.d]: > Deprecation message issued at parse time ? :/

There are depreciation messages issued at lexing as well.

Comment (wilzbach) [reply src/rt/tracegc.d]: Yup, but it looks that this has been fixed already:

https://issues.dlang.org/show_bug.cgi?id=18505

File: src/core/internal/arrayop.d [hunk 2]
```diff
@@ -141,7 +141,7 @@ template typeCheck(bool fail, T, Args...)
     {
         alias UT = Args[idx - 1];
         enum op = Args[idx][1 .. $];
-        static if (is(typeof((UT a) => mixin(op ~ " a")) RT == return))
+        static if (is(typeof((UT a) => mixin(op ~ "cast(int) a")) RT == return))
```
Comment (wilzbach) [inline src/core/internal/arrayop.d]: I could also have used `-preview=intpromote` here, but as this wouldn't be there when any user code tries to use this, I went with the "safe" approach of explicitly using the existing behavior.

**Comment by wilzbach** [comment]
@thewilsonator @JinShil FYI - this is passing and ready ;-)

**Comment by quickfur** [comment]
Heads up: this PR has caused a regression in my code.  Trying to reduce it to minimal case now, will post details once I reduced it.

**Comment by quickfur** [comment]
Regression filed: https://issues.dlang.org/show_bug.cgi?id=19796

This PR broke array ops (`arr[] = -src[];`).

File: src/core/internal/arrayop.d [hunk 3]
```diff
@@ -271,7 +271,7 @@ string scalarExp(Args...)()
         else static if (isUnaryOp(arg))
         {
             auto op = arg[0] == 'u' ? arg[1 .. $] : arg;
-            stack[$ - 1] = op ~ stack[$ - 1];
+            stack[$ - 1] = op ~ "cast(int)"~ stack[$ - 1];
```
Comment (quickfur) [inline src/core/internal/arrayop.d]: This is the culprit. Deleting `"cast(int)"~` fixes the problem.

I think there's a missing check that `Args` contains integral types! The cast causes problems when performing array ops on `double[]`.

**Comment by wilzbach** [comment]
> Regression filed: https://issues.dlang.org/show_bug.cgi?id=19796

Thanks -> https://github.com/dlang/druntime/pull/2550

**Pull request merged**